packages/runtime-core/src/apiOptions.ts

```diff
@@ -155,24 +188,25 @@ export interface LegacyOptions<
   RawBindings,
   D,
   C extends ComputedOptions,
-  M extends MethodOptions
+  M extends MethodOptions,
+  Mixin extends LegacyComponent
 > {
   el?: any
 
   // state
   // Limitation: we cannot expose RawBindings on the `this` context for data
   // since that leads to some sort of circular inference and breaks ThisType
   // for the entire component.
-  data?: D | ((this: ComponentPublicInstance<Props>) => D)
+  data?: D | ((this: CreateComponentPublicInstance<Props>) => D)
   computed?: C
   methods?: M
   watch?: ComponentWatchOptions
   provide?: Data | Function
   inject?: ComponentInjectOptions
 
   // composition
-  mixins?: LegacyComponent[]
-  extends?: LegacyComponent
+  mixins?: Mixin[]
+  extends?: Mixin
```

Comment:
`extends` can't be `Mixin` because if this typing enforces `extends` to be from type declared in the `mixin`

```ts
// works
 mixins: [mixinA, mixinB, mixinC],
 extends: mixinB,

//doesn't work
mixins: [mixinA, mixinB],
extends: mixinC,
```